src/lib/Inference.hs

```diff
@@ -831,28 +841,39 @@ traverseHoles fillHole = (traverseDecl recur, traverseExpr recur, synthPassAtom)
     recur = traverseHoles fillHole
 
 synthDict :: Type -> SynthDictM Atom
-synthDict ty = do
-  (d, bindingInfo) <- getBinding
-  case bindingInfo of
-    LetBound InstanceLet _ -> do
-      block <- buildScoped $ inferToSynth $ instantiateAndCheck ty d
+synthDict ty = case ty of
+  PiTy b arr body -> synthesizeNow <|> introFirst
```

Comment:
Ooh, so this allows us to actually synthesize a "dictionary" of type
`Foo a ?=> Foo b`? Where the default strategy is to solve `Foo b` in a scope where we know `Foo a`?

I'm a bit surprised that the rest of the synthesis code didn't change. I guess the place this gets used is in userspace, when we ask for `(mlift:MonoidLifter b w) ?=>` inside `runAccum`? That's super, super cool. So this essentially reifies the instance-implies-instance logic to userspace.

Reply:
Yes, exactly! Instead of asking for a specific instance, you can now ask for a synthesized "instance lifter", which is what the synthesis pass is the most useful for. This also has a nice correspondence to the Curry-Howard isomorphism: normal synthesis tries to construct a proof for e.g. `Add (n=>m=>a)` given the rules in scope (instances). In that case, proving `Add (m=>a) ?=> Add (n=>m=>a)` can be done either by:
1. Looking up the proof in the scope
2. Adding `Add (m=>a)` to your assumptions and proceeding as usual! This is why Coq proofs usually start with `intros`